### DISCUSSION TITLE
Is there a way to execute command within a pod using Openshift Python Client?

### DISCUSSION THREAD

**Discussion by LiquIDMeowz** [openshift/openshift-client-python#137] [Q&A] [answered]
Hi, 

I'm trying to get ETCD Database size using the Python Client for OpenShift. 
After authenticating against OpenShift, I can get the pod by name (for example v1_pod = dyn_client.resources.get(api_version='v1', kind='Pod', name='some_pod_name'), but how do I execute a command in the pod using the client? 

I tried using subprocess.run, however due to a missing OC client on node side, the command fails. 
My script looks like this and I'll create an image from it and run it as a daily CronJob:

`from kubernetes import client, config, dynamic
from openshift.dynamic import DynamicClient
from openshift.helper.userpassauth import OCPLoginConfiguration
import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
import urllib3
urllib3.disable_warnings()
import subprocess

### Mail template
def send_mail(etcd_db):
    mail_message = f"ETCD DB Sizes of {cname} Cluster -- {proc.stdout}"
    msg = MIMEMultipart("alternative")
    msg['From'] = "<redacted>"
    msg['To'] = "<redacted>"
    msg['Subject'] = f"ETCD DB Size for Cluster {cname}"
    message_render = MIMEText(mail_message, "html")
    msg.attach(message_render)
    s = smtplib.SMTP('smtp.svcs.entsvcs.com', 25)
    s.send_message(msg)

### Authenticate against OCP
with open('/var/run/secrets/kubernetes.io/serviceaccount/token', 'r') as f:
  token = f.readline().strip()
token_auth = dict(
    api_key={'authorization': 'Bearer {}'.format(token)},
    host='<redacted>',
    verify_ssl=False
)
configuration = client.Configuration()
for k, v in token_auth.items():
    setattr(configuration, k, v)
k8s_client = client.ApiClient(configuration)
dyn_client = DynamicClient(k8s_client)

### Get cluster name and set cname to = <redacted>/<redacted>/<redacted>
v1_name = dyn_client.resources.get(api_version='v1', kind='ConfigMap')
try:
  cluster = v1_name.get(name='<redacted>', namespace="<redacted>").data.worker
  cname = "<redacted>"
except:
  try:
    cluster = v1_name.get(name='<redacted>', namespace="<redacted>").data.worker
    cname = "<redacted>"
  except:
    cluster = v1_name.get(name='<redacted>', namespace="<redacted>").data.worker
    cname = "<redacted>"

### Check DB Size and send E-mail
etcd_db = "oc exec <redacted> -c etcdctl -n openshift-etcd -- etcdctl endpoint status -w table --cluster | awk '{print $8, $9}' | tail -n 4"
proc = subprocess.run(etcd_db, shell=True, encoding='utf-8', stdout=subprocess.PIPE)
print(cname)
print(proc.stdout)
send_mail(etcd_db)`

Thank you!
LiquID

**Comment by bradmwilliams** [comment, marked as answer]
Hi @LiquIDMeowz,
Here is a script that should provide you a starting point for launching commands on a pod:
```python
#!/usr/bin/python

import openshift as oc

if __name__ == '__main__':
    options = {
        'as': 'system:admin',
        'context': 'my_cluster_1',
    }

    with oc.client_host():
        with oc.timeout(60 * 5):
            with oc.options(options):
                with oc.project("openshift-etcd"):
                    pods = oc.selector("pods", labels={'app': 'etcd'}).objects()
                    print(f'Found: {len(pods)} pods')
                    result = pods[0].execute(cmd_to_exec=['etcdctl', 'endpoint', 'status', '--cluster', '-w', 'table'])
                    print(f'Result:\n{result.out()}')

```

**Reply by bradmwilliams** [reply]
Please note, based on your question, it sounds like you might actually be using the `openshift-restclient-python` library and the library associated with this repo `openshift-client-python`.  They are not the same.  My answer above would accomplish what you're looking for, but uses the `openshift-client-python` library.

**Comment by LiquIDMeowz** [comment]
@bradmwilliams Hi, 

Thank you very much!
